README.md
<img src="https://github.com/dmslabsbr/hoymiles/raw/master/logo.png" alt="" width="200" />


# HoyMiles Solar Data Gateway Add-on
Application to read Hoymiles Gateway Solar Data

I developed this program to integrade my solar system data to the [Home Assistant](https://www.home-assistant.io/) Application through an add-on.

But you can use the application without using the Home Assistant. You just need a machine that runs Python3.

<a href="https://www.buymeacoffee.com/dmslabs"><img src="https://img.buymeacoffee.com/button-api/?text=Buy me a pizza&emoji=🍕&slug=dmslabs&button_colour=FFDD00&font_colour=000000&font_family=Cookie&outline_colour=000000&coffee_colour=ffffff"></a>

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=9S3JYKPHR3XQ6)
[![Donate with Bitcoin](https://en.cryptobadges.io/badge/micro/1MAC9RBnPYT9ua1zsgvhwfRoASTBKr4QL8)](https://www.blockchain.com/btc/address/1MAC9RBnPYT9ua1zsgvhwfRoASTBKr4QL8)

<img alt="Lines of code" src="https://img.shields.io/tokei/lines/github/dmslabsbr/hoymiles">
<img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/dmslabsbr/hoymiles">


# Instructions

<img align="center" src="https://github.com/dmslabsbr/smsUps/raw/master/hass.io.png" alt="" width="30" /> [Home Assistant add-on instructions](DOCS.md)


Before run you need to install:
   https://github.com/eclipse/paho.mqtt.python  ***and***
   https://github.com/psf/requests


```bash
git clone https://github.com/dmslabsbr/hoymiles.git
cd hoymiles
python3 -m venv ./hoymiles/
source ./bin/activate
pip3 install paho-mqtt
pip3 install requests
```

My solar panels communicate with the internet using a DTU-W100 gateway.

<img src="https://github.com/dmslabsbr/hoymiles/raw/master/icon.png" alt="" width="300" />

But it will probably work with any device that uses the [global.hoymiles.com](https://global.hoymiles.com/) Website


## PS:
I invite everyone to help in the this tool development.

## Screenshots

<img src="https://github.com/dmslabsbr/hoymiles/blob/master/img/Hass1.png?raw=true" alt="" width="400" />

<img src="https://github.com/dmslabsbr/hoymiles/blob/master/img/Hass2.png?raw=true" alt="" width="400" />

<img src="https://github.com/dmslabsbr/hoymiles/blob/master/img/Hass3.png?raw=true" alt="" width="400" />



#### Licence

> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# CHANGELOG

## 0.24

        - Refactor of whole code
        - Optimization of whole flow.
        - webserver is disabled
        - switch to read config from json
        - changed sensor reset date to the begining of next day

## 0.23

        - Correct "Missing default value for External_MQTT_TLS bug' - issue #44
        - Try to solve issue #40 - 'dict object' has no attribute

## 0.22

        - Correct - Option 'MQTT_HOST' does not exist in the schema for HoyMiles Solar Data Gateway Add-on
        - Add device_class: power and state_class: measurement to real_power template
        - TLS external server test
        - Add timezone to last_data_time - issue #42

## 0.21

        - External MQTT bug fix
  
        - New sensors magnitude

### 0.20

"0.20i" - Support to the HA Energy Management dashboard (BETA)

        - Change some sensors config (testing)
  
        - Automatically receives login data from the mqqt server and ignores the config.
  
        - Load HASS.IO MQTT host, user and pass
  
        - Dark mode bug fix

### 0.19

"0.19" - amd64-base-python:3.9-alpine3.12 bug - now using 3.7-alpine3.12

### 0.18

"0.18" - fix returned a non-zero code on install
       Alpine 3.13 bug - now using 3.12

### 0.17

Dockerfile error

### 0.16

- First webserver version
  
- Comum.py
  
- https://www.buymeacoffee.com/dmslabs
  
- Images

### 0.15

- Bug fix
  
- dic['expire_after'] error

### 0.14
- float2number fix

### 0.13
- Bug fix

### 0.12
- Realpower round
  
- Other data round and scale
  
- sensors update

### 0.11
- Bug fix

### 0.10
- MQTT_STATUS_CODE
  
- Sensor DIC
  
- More debug information
  
- dmslibs.PrintC

### 0.09

- bug fix pegaDadosSolar()
  
- fixed I can't get access token
  
- changed loop to get data

### 0.08 

- DEVELOPERS_MODE logs
  
- Changer Network port in config.json

### 0.07 

- Add DEVELOPERS_MODE at add-on config

### 0.06 

- Power ratio * 100;
  
- Expire interval = Get Data Interval * 1.5


const.py
"""
Module contain consts and templates used in addon.
"""

import datetime

HTTP_STATUS_CODE = {
    100: "Continue",
    200: "OK",
    202: "Non-Authoritative Information",
    307: "Temporary Redirect",
    308: "Permanent Redirect",
    400: "Bad Request",
    401: "Unauthorized",
    403: "Forbidden",
    404: "Not Found",
    407: "Proxy Authentication Required",
    408: "Request Timeout",
    500: "Internal Server Error",
    502: "Bad Gateway",
    504: "Gateway Timeout",
    1000: "Message not set."
}

MQTT_STATUS_CODE = {
    0: "Connection successful",
    1: "Connection refused – incorrect protocol version",
    2: "Connection refused – invalid client identifier",
    3: "Connection refused – server unavailable",
    4: "Connection refused – bad username or password",
    5: "Connection refused – not authorised",
    100: "Connection refused - other things"
}

PAYLOAD_T1 = '''
   {
       "ERROR_BACK":true,
       "LOAD":{
           "loading":true
        },
        "body":{
            "password":"$password",
            "user_name":"$user"
        },
        "WAITING_PROMISE":true
    }
'''

PAYLOAD_T2 = '''
{
    "body": {
        "sid": $sid
    },
    "WAITING_PROMISE": true
}
'''

PAYLOAD_ID = '''
{
    "body": {
        "id": $id
    },
    "WAITING_PROMISE": true
}
'''

HEADER_LOGIN = {
    'Content-Type': 'application/json;charset=UTF-8',
    'Cookie': ''
}

HEADER_DATA = {
    'Content-Type': 'application/json;charset=UTF-8',
    'Cache-Control': 'no-cache',
    'Host': 'global.hoymiles.com',
    'Connection': 'keep-alive',
    'Accept': 'application/json, text/plain, */*',
    'Accept-Encoding': 'gzip, deflate, br',
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36',
    'Accept-Language': 'pt-BR,pt;q=0.9,it-IT;q=0.8,it;q=0.7,es-ES;q=0.6,es;q=0.5,en-US;q=0.4,en;q=0.3',
    'Cookie': 'hm_token_language=en_us; '
}


SECRETS = 'secrets.ini'

COOKIE_UID = "'uid=fff9c382-389f-4a47-8dc9-c5486fc3d9f5"
COOKIE_EGG_SESS = "EGG_SESS=XHfAhiHWwU__OUVeKh0IiITBnmwA-IIXEzTCHgHgww6ZYYddOPntPSwVz4Gx7ISbfU0WrvzOLungThcL-9D2KxavrtyPk8Mr2YXLFzJwvM0usPvhzYdt2Y2S9Akt5sjP'"


BASE_URL = "https://global.hoymiles.com/platform/api/gateway/"
LOGIN_API = "iam/auth_login"
USER_ME = "iam/user_me"
GET_DATA_API = "pvm-data/data_count_station_real_data"
GET_ALL_DEVICE_API = "pvm/station_select_device_all"
STATION_FIND = "pvm/station_find"

# For MQTT
MQTT_PUB = "home/solar"
SID = 'solar'
MQTT_HASS = "homeassistant"
DEFAULT_MQTT_PASS = "MQTT_PASSWORD"
NODE_ID = 'dmslabs'
SHORT_NAME = 'solarH'
HASS_INTERVAL = 1200
GETDATA_INTERVAL = 480  # How often do I read site data
EXPIRE_TIME = int(GETDATA_INTERVAL) * 1.5


LOCAL_TIMEZONE = datetime.datetime.now(
    datetime.timezone.utc).astimezone().tzinfo


json_hass = {"sensor": '''
{ 
  "stat_t": "home/$sid/json_$via_device",
  "name": "$name",
  "uniq_id": "$uniq_id",
  "val_tpl": "{{ value_json.$val_tpl }}",
  "icon": "$icon",
  "device_class": "$device_class",
  "state_class": "$state_class",
  "unit_of_measurement": "$unit_of_measurement",
  "expire_after": "$expire_after",
  "device": { $device_dict }
}''',
             "binary_sensor": '''
{ 
  "stat_t": "home/$sid/json_$via_device",
  "name": "$name",
  "uniq_id": "$uniq_id",
  "val_tpl": "{{ value_json.$val_tpl }}",
  "icon": "$icon",
  "device_class": "$device_class",
  "device": { $device_dict }
}'''
}

DEVICE_DICT = ''' "name": "$device_name",
    "manufacturer": "$manufacturer",
    "model": "$model",
    "sw_version": "$sw_version",
    "via_device": "$via_device",
    "identifiers": [ "$identifiers" ] '''
